unikernel.ml

```diff
@@ -800,75 +911,173 @@ struct
-                            in
-                            (* TODO: middleware, extract domain from middleware *)
-                            Albatross.query !albatross ~domain:"robur" ~name
-                              (`Unikernel_cmd (`Unikernel_create config))
-                            >|= function
-                            | Error () ->
-                                Logs.warn (fun m ->
-                                    m "error querying albatross");
-                                reply "error while querying albatross"
-                            | Ok None ->
-                                Logs.warn (fun m -> m "got none");
-                                reply "got none"
-                            | Ok (Some (_hdr, res)) ->
-                                reply_json (Albatross_json.res res))
+        | "/unikernel/create" -> (
+            let request = Httpaf.Reqd.request reqd in
+            match request.meth with
+            | `POST ->
+                let now = Ptime.v (P.now_d_ps ()) in
+                let _, (t : Storage.t) = !store in
+                let users = User_model.create_user_session_map t.users in
+                let middlewares = [ Middleware.auth_middleware now users ] in
+                Middleware.apply_middleware middlewares
+                  (fun _reqd ->
+                    let response_body = Httpaf.Reqd.request_body reqd in
+                    let finished, notify_finished = Lwt.wait () in
+                    let wakeup v = Lwt.wakeup_later notify_finished v in
+                    let on_eof data () = wakeup data in
+                    let f acc s = acc ^ s in
+                    let rec on_read on_eof acc bs ~off ~len =
+                      let str = Bigstringaf.substring ~off ~len bs in
+                      let acc = acc >>= fun acc -> Lwt.return (f acc str) in
+                      Httpaf.Body.schedule_read response_body
+                        ~on_read:(on_read on_eof acc) ~on_eof:(on_eof acc)
+                    in
+                    let f_init = Lwt.return "" in
+                    Httpaf.Body.schedule_read response_body
+                      ~on_read:(on_read on_eof f_init) ~on_eof:(on_eof f_init);
+                    finished >>= fun data ->
+                    data >>= fun data ->
+                    let content_type =
+                      Httpaf.(
+                        Headers.get_exn (Reqd.request reqd).Request.headers
+                          "content-type")
+                    in
+                    let ct =
+                      Multipart_form.Content_type.of_string
+                        (content_type ^ "\r\n")
+                    in
+                    match ct with
+                    | Error (`Msg msg) ->
+                        Logs.warn (fun m -> m "couldn't content-type: %s" msg);
+                        let status =
+                          {
+                            Utils.Status.code = 403;
```

Comment:
is 403 the right one here? isn't it rather 400 - bad request?

Reply:
I tend to agree - the client sent malformed request body